### PR TITLE
tests: integration: device_test: modify kw device integration test to run entirely in container

### PR DESCRIPTION
**Pull request by aquilamacedo**
This PR modifies the `device_test` integration test to run entirely within a container. The changes include restructuring the test to use a results array to store the output of the `kw device --local` command from Debian, Fedora, and Arch distributions. New helper functions were added to handle the setup and execution of tests for RAM, storage, OS, motherboard, chassis, CPU, and GPU information. Minor adjustments were made to escape single quotes in the `container_exec` function to ensure proper command execution.

### CODE REVIEW

File: tests/integration/device_test.sh [hunk 1]
```diff
@@ -16,67 +16,39 @@ function device_info_test_helper()
...
 
-  # add | separator to each filter item.
-  filter_regex=$(sed 's/ /|/g' <<< "${filter[@]}")
+    # Copy the kw_device_expected.sh to the container's workdir directory
+    container_copy "$container" "$kw_device_simulation" "$container_tmp_dir"
 
-  # the actual perl regex.
-  filter_regex="(${filter_regex}):"
+    # Get the output of the `kw device --local` command from the container
+    buffer=$(container_exec "$container" 'kw device --local')
```
Comment (rodrigosiqueira) [inline tests/integration/device_test.sh:31]: maybe rename this buffer variable to something like expected_output_container.

File: tests/integration/device_test.sh [hunk 2]
```diff
@@ -16,67 +16,39 @@ function device_info_test_helper()
...
 
-  # get the output, after applying filter.
-  output=$(printf '%s' "${buffer}" | grep --invert-match --perl-regexp "${filter_regex}")
+    # Specify the fields to filter out from the output
```
Comment (rodrigosiqueira) [inline tests/integration/device_test.sh:33]: Consider adding a sample example of the type of output that this filter will work on.

File: tests/integration/device_test.sh [hunk 3]
```diff
@@ -16,67 +16,39 @@ function device_info_test_helper()
...
-  assertEquals "(${LINENO}): kw device failed for ${distro}" "${expected_output}" "${output}"
+    # Get the output, after applying filter.
+    output=$(printf '%s' "${buffer}" | grep --invert-match --perl-regexp "$filter_regex")
 
-  # check storage information
-  fs_type=$(container_inspect --format '{{.Driver}}' "${container}")
-  fs_size=$(container_inspect --format '{{.GraphDriver.Data.WorkDir}}' "${container}" |
-    xargs df -h | tail --lines 1 | awk '{ print $2 }')
+    # Get the expected output
+    expected_output=$(container_exec "$container" "bash ${container_tmp_dir}/kw_device_expected_output.sh")
```
Comment (rodrigosiqueira) [inline tests/integration/device_test.sh:48]: tbh, I did not like this approach since it creates a script similar to the device that we have to maintain. How about changing how we approach this problem a little bit? We can try something like this:
1. Run kw device in the container and capture the output.
2. Create many small tests that check just part of the output. 

For example, you can create one test just for validate the gpu return; for that, you just run the command to retrive the gpu output from the container and compare it with the result from `kw device`.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

**Comment by aquilamacedo** [comment]
Hi @davidbtadokoro and @rodrigosiqueira , I made changes based on our conversation at the last meeting, could you check if everything is in order?

File: tests/integration/device_test.sh [hunk 4]
```diff
@@ -4,87 +4,261 @@ include './tests/unit/utils.sh'
...
-  buffer=$(container_exec "${container}" 'kw device --local')
+  for distro in "${!results[@]}"; do
+    container="kw-${distro}"
+    output="${results[$distro]}"
+    # Get the RAM value in kilobytes from the container
+    mem_total_kb=$(container_exec "$container" "awk '/MemTotal/ {print \$2}' /proc/meminfo")
+    # Convert the RAM value to IEC format
+    expected_output=$(container_exec "$container" "numfmt --from=si --to=iec ${mem_total_kb}K")
+    # Extract RAM value from the output
+    output=$(printf "%s" "$output" | grep "Total RAM" | awk '{print $3}')
```
Comment (rodrigosiqueira) [inline tests/integration/device_test.sh:65]: awk in fine, but maybe a simple `cut -d` can get the job done and it is very easy to understand.

File: tests/integration/device_test.sh [hunk 5]
```diff
@@ -4,87 +4,261 @@ include './tests/unit/utils.sh'
...
+    # Extract actual root filesystem from output
+    output_fs=$(printf "%s" "${results[$distro]}" | awk '/Root filesystem:/ {print $3}')
+    # Assert that expected and actual root filesystem values match
+    assert_equals_helper "'kw device' Root filesystem mismatch for ${distro}" "$LINENO" "$expected_fs" "$output_fs"
+
+    # Retrieve expected root filesystem size from container
+    expected_size=$(container_exec "$container" "df -h / | tail -n 1 | awk '{print \$2}'")
+    # Extract actual root filesystem size from output
+    output_size=$(printf "%s" "${results[$distro]}" | awk '/Size:/ {print $2}')
+    # Assert that expected and actual root filesystem size values match
```
Comment (rodrigosiqueira) [inline tests/integration/device_test.sh:96]: I'm under the impression that all of these comments could be used to create a better variable name and, with that, get rid of the comments.

File: tests/integration/device_test.sh [hunk 6]
```diff
@@ -4,87 +4,261 @@ include './tests/unit/utils.sh'
...
+  local distro
+  local container
+  local expected_distro
+  local output_distro
+
+  for distro in "${!results[@]}"; do
+    container="kw-${distro}"
+
+    # Extract expected distribution name from the container's /etc/os-release
+    expected_distro=$(container_exec "$container" "grep ^NAME= /etc/os-release | cut -d'\"' -f2")
```
Comment (rodrigosiqueira) [inline tests/integration/device_test.sh:119]: Please, for this entire commit, review:
https://kworkflow.org/content/codingstyle.html#avoid-short-options

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]
Overall, this new version looks really good. Check my comments and expand them to your entire PR.

**Comment by aquilamacedo** [comment]
Hi @rodrigosiqueira. I made the changes based on your review, could you check if everything is in order?

**Review by davidbtadokoro — CHANGES_REQUESTED** [review]
Hi @aquilamacedo, thank you so much tackling this long known issue :smiling_face_with_tear:

Overall, it looks legit! Both the isolation and the extraction into specific tests are great :+1: 

I've tested it and had a fail on my end, but fear not as I really think is just an odd edge case you didn't envisioned, and that is safe to tackle. Besides that, it seems that this PR really will solve this looooong-lasting issue.

Don't know if I am confusing things, but if there is an issue for this problem, don't forget to add the `Closes:` tag at the end of the second commit.

File: tests/integration/device_test.sh [hunk 7]
```diff
@@ -4,87 +4,223 @@ include './tests/unit/utils.sh'
...
+declare -gA DEVICE_INFO_RESULTS
+
 function oneTimeSetUp()
 {
   # The VERBOSE variable is set and exported in the run_tests.sh script based
   # on the command-line options provided by the user. It controls the verbosity
   # of the output during the test runs.
   setup_container_environment "$VERBOSE" 'device'
+  # Ensure kw_device_output is called once to populate the results array
+  kw_device_output
```
Comment (rodrigosiqueira) [inline tests/integration/device_test.sh:16]: There is no need to create this function since you just call it once. Just move the content of this function to the oneTimeSetUp.

File: tests/integration/device_test.sh [hunk 8]
```diff
@@ -4,87 +4,223 @@ include './tests/unit/utils.sh'
...
+    assert_equals_helper "'kw device' CPU max frequency mismatch for ${distro}" "$LINENO" "$expected_cpu_max_freq" "$output_cpu_max_freq"
+
+    expected_cpu_min_freq=$(container_exec "$container" "lscpu | grep 'CPU min MHz' | cut --delimiter ':' --fields 2 | xargs")
+    output_cpu_min_freq=$(printf "%s" "${DEVICE_INFO_RESULTS[$distro]}" | grep 'Min frequency (MHz):' | cut --delimiter ':' --fields 2 | xargs)
+    assert_equals_helper "'kw device' CPU min frequency mismatch for ${distro}" "$LINENO" "$expected_cpu_min_freq" "$output_cpu_min_freq"
+  done
+}
+
+# Test the GPU information reported by 'kw device --local'
+function test_kw_device_local_gpu()
```
Comment (rodrigosiqueira) [inline tests/integration/device_test.sh:182]: This function still has failures in the local run:

```
test_kw_device_local_gpu
ASSERT: line 215: 'kw device' GPU model mismatch for archlinux at 0c:00.0
  Expected Result: Model: Advanced Micro Devices, Inc. [AMD/ATI] Navi 14 [Radeon RX 5500/5500M / Pro 5500M]
    Actual Result: GPU:
  Model: Advanced Micro Devices, Inc. [AMD/ATI] Navi 14 [Radeon RX 5500/5500M / Pro 5500M]
  Provider: Advanced Micro Devices, Inc. 
ASSERT: line 215: 'kw device' GPU model mismatch for debian at 0c:00.0
  Expected Result: Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
    Actual Result: GPU:
  Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
  Provider: Advanced Micro Devices, Inc. 
ASSERT: line 215: 'kw device' GPU model mismatch for fedora at 0c:00.0
  Expected Result: Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
    Actual Result: GPU:
  Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
  Provider: Advanced Micro Devices, Inc. 

Ran 7 tests.
```

File: tests/integration/device_test.sh [hunk 9]
```diff
@@ -4,87 +4,223 @@ include './tests/unit/utils.sh'
...
+  local expected_gpu_provider
+  local output_gpu_model
+  local output_gpu_provider
+  declare -A expected_gpus
+
+  for distro in "${DISTROS[@]}"; do
+    container="kw-${distro}"
+
+    pci_addresses=$(container_exec "$container" "lspci | grep --regexp=VGA --regexp=Display --regexp=3D | cut --delimiter=' ' --fields 1")
+    for g in $pci_addresses; do
```
Comment (rodrigosiqueira) [inline tests/integration/device_test.sh:198]: Please use a better variable name here; `g` is not good.

File: tests/integration/device_test.sh [hunk 10]
```diff
@@ -4,87 +4,223 @@ include './tests/unit/utils.sh'
...
+    # Capture all lines starting from 'GPU:' until the next empty line
+    output_gpu=$(printf "%s" "${DEVICE_INFO_RESULTS[$distro]}" | sed -n '/^GPU:/,/^$/p')
+
+    for g in "${!expected_gpus[@]}"; do
+      expected_gpu_model="Model: $(printf '%s' "${expected_gpus[$g]}" | cut --delimiter ';' --fields 1)"
+      expected_gpu_provider="Provider: $(printf '%s' "${expected_gpus[$g]}" | cut --delimiter ';' --fields 2)"
+
+      # Check GPU model
+      printf "%s" "$output_gpu" | grep -q "$expected_gpu_model"
+      if [ $? -ne 0 ]; then
```
Comment (rodrigosiqueira) [inline tests/integration/device_test.sh:214]: Please don't use this style of `if`; this is not the default to the project. We prefer `[[ ]]`.

File: tests/integration/device_test.sh [hunk 11]
```diff
@@ -4,87 +4,223 @@ include './tests/unit/utils.sh'
...
+
+    # Capture all lines starting from 'GPU:' until the next empty line
+    output_gpu=$(printf "%s" "${DEVICE_INFO_RESULTS[$distro]}" | sed -n '/^GPU:/,/^$/p')
+
+    for g in "${!expected_gpus[@]}"; do
+      expected_gpu_model="Model: $(printf '%s' "${expected_gpus[$g]}" | cut --delimiter ';' --fields 1)"
+      expected_gpu_provider="Provider: $(printf '%s' "${expected_gpus[$g]}" | cut --delimiter ';' --fields 2)"
+
+      # Check GPU model
+      printf "%s" "$output_gpu" | grep -q "$expected_gpu_model"
```
Comment (rodrigosiqueira) [inline tests/integration/device_test.sh:213]: https://kworkflow.org/content/codingstyle.html#avoid-short-options

Comment (rodrigosiqueira) [reply tests/integration/device_test.sh:213]: I debugged the failure issue a bit, and I concluded that the above grep is wrong. Check this extra log that I added for debug:

```
test_kw_device_local_gpu
>>>>>
Model: Advanced Micro Devices, Inc. [AMD/ATI] Navi 14 [Radeon RX 5500/5500M / Pro 5500M]
>>>>>
GPU:
  Model: Advanced Micro Devices, Inc. [AMD/ATI] Navi 14 [Radeon RX 5500/5500M / Pro 5500M]
  Provider: Advanced Micro Devices, Inc. 
>>>>
~~> 1
ASSERT: line 224: 'kw device' GPU model mismatch for archlinux at 0c:00.0
  Expected Result: Model: Advanced Micro Devices, Inc. [AMD/ATI] Navi 14 [Radeon RX 5500/5500M / Pro 5500M]
    Actual Result: GPU:
  Model: Advanced Micro Devices, Inc. [AMD/ATI] Navi 14 [Radeon RX 5500/5500M / Pro 5500M]
  Provider: Advanced Micro Devices, Inc. 
>>>>>
Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
>>>>>
GPU:
  Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
  Provider: Advanced Micro Devices, Inc. 
>>>>
~~> 1
ASSERT: line 224: 'kw device' GPU model mismatch for debian at 0c:00.0
  Expected Result: Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
    Actual Result: GPU:
  Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
  Provider: Advanced Micro Devices, Inc. 
>>>>>
Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
>>>>>
GPU:
  Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
  Provider: Advanced Micro Devices, Inc. 
>>>>
~~> 1
ASSERT: line 224: 'kw device' GPU model mismatch for fedora at 0c:00.0
  Expected Result: Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
    Actual Result: GPU:
  Model: Advanced Micro Devices, Inc. [AMD/ATI] Device 0b0c
  Provider: Advanced Micro Devices, Inc.
```
In particular, notice:

```
Actual Result: GPU:
```

Which indicates that the grep that you are trying to use is wrong. I tried a quick workaround for this:

```
printf "%s" "$output_gpu" | grep -q "Model:"
```

The above work around fixed the issue. Please investigate more and find a better implementation of my suggestion.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

**Comment by rodrigosiqueira** [comment]
Thanks, change merged into the unstable branch.
